cosmwasm/contracts/ibc-translator/src/reply.rs

```diff
@@ -0,0 +1,174 @@
+        .context("no data in the response, we should never get here")?;
+    let res_data: CompleteTransferResponse =
+        from_binary(&res_data_raw).context("failed to deserialize response data")?;
+    let contract_addr = res_data
+        .contract
+        .context("no contract in response, we should never get here")?;
+
+    // load interim state
+    let transfer_info = CURRENT_TRANSFER
+        .load(deps.storage)
+        .context("failed to load current transfer from storage")?;
+
+    // delete interim state
+    CURRENT_TRANSFER.remove(deps.storage);
+
+    // deserialize payload into the type we expect
+    let payload: GatewayIbcTokenBridgePayload = serde_json_wasm::from_slice(&transfer_info.payload)
+        .context("failed to deserialize transfer payload")?;
+
+    match payload {
+        GatewayIbcTokenBridgePayload::Simple {
+            chain,
+            recipient,
+            fee: _,
+            nonce: _,
+        } => {
+            let recipient_decoded = String::from_utf8(recipient.to_vec())
+                .context(format!("failed to convert {recipient} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                recipient_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                None,
+            )
+        }
+        GatewayIbcTokenBridgePayload::ContractControlled {
+            chain,
+            contract,
+            payload,
+            nonce: _,
+        } => {
+            let contract_decoded = String::from_utf8(contract.to_vec())
+                .context(format!("failed to convert {contract} to utf8 string"))?;
+            convert_cw20_to_bank_and_send(
+                deps,
+                env,
+                contract_decoded,
+                res_data.amount.into(),
+                contract_addr,
+                chain,
+                Some(payload),
+            )
+        }
+    }
+}
+pub fn convert_cw20_to_bank_and_send(
+    deps: DepsMut,
+    env: Env,
+    recipient: String,
+    amount: u128,
+    cw20_contract_addr: String,
+    chain_id: u16,
+    payload: Option<Binary>,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    deps.api
+        .addr_validate(&cw20_contract_addr)
+        .context(format!("invalid contract address {cw20_contract_addr}"))?;
+
+    // convert contract address into base64
+    let subdenom = contract_addr_to_base58(deps.as_ref(), cw20_contract_addr.clone())?;
+    // format the token factory denom
+    let tokenfactory_denom = "factory/".to_string()
+        + env.contract.address.to_string().as_ref()
+        + "/"
+        + subdenom.as_ref();
+
+    let mut response: Response<TokenFactoryMsg> = Response::new();
+
+    // check contract storage see if we've created a denom for this cw20 token yet
+    // if we haven't created the denom, then create the denom
+    // info.sender contains the cw20 contract address
```

Comment:
What does this comment mean exactly? I tried to find context for it somewhere but I failed.

Reply:
Good catch, we can remove this comment.